cctp_primitives/src/commitment_tree/hashers.rs

```diff
@@ -1,152 +1,233 @@
-                 sc_request_data: &[u8],
-                 pk_hash: &[u8],
-                 tx_hash: &[u8],
-                 out_idx: u32)
-    -> Result<FieldElement, Error> {
-    let mut bytes = Vec::<u8>::new();
-
-    bytes.write_i64::<BigEndian>(sc_fee)?;
-    bytes.extend(&sc_request_data.to_vec());
-    bytes.extend(&pk_hash.to_vec());
-    bytes.extend(&tx_hash.to_vec());
-    bytes.write_u32::<BigEndian>(out_idx)?;
-
-    hash_bytes(&bytes)
-}
+pub fn hash_bwtr(
+    sc_fee:  u64,
+    sc_request_data: &[[u8; FIELD_SIZE]],
+    mc_destination_address: &[u8; 20],
+    tx_hash: &[u8; 32],
+    out_idx: u32
+) -> Result<FieldElement, Error>
+{
+    let mut buffer = Vec::new();
+    sc_fee.write(&mut buffer)?;
+    mc_destination_address.write(&mut buffer)?;
+    tx_hash.write(&mut buffer)?;
+    out_idx.write(&mut buffer)?;
 
-// Computes FieldElement-based hash on the given Certificate data
-pub fn hash_cert(epoch_number: u32,
-                 quality: u64,
-                 cert_data_hash: &[u8],
-                 bt_list: &[(i64,[u8; 20])],
-                 custom_fields_merkle_root: &[u8],
-                 end_cumulative_sc_tx_commitment_tree_root: &[u8])
-    -> Result<FieldElement, Error> {
-    let mut bytes = Vec::<u8>::new();
-
-    bytes.write_u32::<BigEndian>(epoch_number)?;
-    bytes.write_u64::<BigEndian>(quality)?;
-    bytes.extend(&cert_data_hash.to_vec());
-    bytes.extend(&bt_list_to_bytes(bt_list)?);
-    bytes.extend(&custom_fields_merkle_root.to_vec());
-    bytes.extend(&end_cumulative_sc_tx_commitment_tree_root.to_vec());
-
-    hash_bytes(&bytes)
-}
+    // ceil(256 + 160 + 96/254) = ceil(512/254) = 3 fes
+    let mut fes = bytes_to_field_elements(buffer)?;
 
-// Computes FieldElement-based hash on the given Sidechain Creation Transaction data
-pub fn hash_scc(amount: i64,
-                pub_key: &[u8],
-                withdrawal_epoch_length: u32,
-                custom_data: &[u8],
-                constant: Option<&[u8]>,
-                cert_verification_key: &[u8],
-                btr_verification_key: Option<&[u8]>,
-                csw_verification_key: Option<&[u8]>,
-                tx_hash: &[u8],
-                out_idx: u32)
-    -> Result<FieldElement, Error> {
-    let mut bytes = Vec::<u8>::new();
-
-    bytes.write_i64::<BigEndian>(amount)?;
-    bytes.extend(&pub_key.to_vec());
-    bytes.write_u32::<BigEndian>(withdrawal_epoch_length)?;
-    bytes.extend(&custom_data.to_vec());
-    if constant.is_some(){
-        bytes.extend(&constant.unwrap().to_vec());
+    // sc_request_data elements MUST BE field elements
+    for fe in sc_request_data.iter() {
+        fes.push(FieldElement::read(&fe[..])?);
     }
-    bytes.extend(&cert_verification_key.to_vec());
-    if btr_verification_key.is_some(){
-        bytes.extend(&btr_verification_key.unwrap().to_vec());
+
+    Ok(hash_vec(fes))
+}
+
+pub fn get_cert_data_hash(
+    constant: Option<&[u8; FIELD_SIZE]>,
```

Comment:
Is it really necessary to include `constant` in a cert data hash, i.g. in each Certificate?
From SC perspective, if `constant` is a part of the CommitmentTree -> we should keep it as a part of the Certificate (because SC knows only about its own `constant` and doesn't track other SCs data), but the constant is always the same for a specific sidechain.

Reply:
It seems reasonable, unless there are other reasons ( @albertog78 , @dkaidalov ) ?

Reply:
remove `constant` from the `cert_data_hash` and `add_cert` methods.